src/routers/account.js

```diff
@@ -494,4 +499,150 @@ router.get('/miieditor', async (request, response) => {
+router.post('/checkout/:priceId', async (request, response) => {
+	// Verify the user is logged in
+	if (!request.cookies.access_token || !request.cookies.refresh_token || !request.cookies.ph) {
+		return response.redirect('/account/login');
+	}
+
+	// Attempt to get user data
+	let apiResponse = await util.apiGetRequest('/v1/user', {
+		'Authorization': `${request.cookies.token_type} ${request.cookies.access_token}`
+	});
+
+	if (apiResponse.statusCode !== 200) {
+		// Assume expired, refresh and retry request
+		apiResponse = await util.apiPostGetRequest('/v1/login', {}, {
+			refresh_token: request.cookies.refresh_token,
+			grant_type: 'refresh_token'
+		});
+
+		if (apiResponse.statusCode !== 200) {
+			return response.redirect('/account/login');
+		}
+
+		const tokens = apiResponse.body;
+
+		response.cookie('refresh_token', tokens.refresh_token, { domain: '.pretendo.network' });
+		response.cookie('access_token', tokens.access_token, { domain: '.pretendo.network' });
+		response.cookie('token_type', tokens.token_type, { domain: '.pretendo.network' });
+
+		apiResponse = await util.apiGetRequest('/v1/user', {
+			'Authorization': `${tokens.token_type} ${tokens.access_token}`
+		});
+	}
+
+	// If still failed, something went horribly wrong
+	if (apiResponse.statusCode !== 200) {
+		return response.redirect('/account/login');
+	}
+
+	// Set user account info to render data
+	const account = apiResponse.body;
+	const pid = account.pid;
+
+	let customer;
+	const { data: searchResults } = await await stripe.customers.search({
+		query: `metadata['pnid_pid']:'${pid}'`
+	});
+
+	if (searchResults.length !== 0) {
+		customer = searchResults[0];
+	} else {
+		customer = await stripe.customers.create({
+			metadata: {
+				pnid_pid: pid
+			}
+		});
+	}
+
+	const priceId = request.params.priceId;
+
+	const session = await stripe.checkout.sessions.create({
+		line_items: [
+			{
+				price: priceId,
+				quantity: 1,
+			},
+		],
+		customer: customer.id,
+		mode: 'subscription',
+		success_url: `${config.http.base_url}/account?upgrade_success=true`,
+		cancel_url: `${config.http.base_url}/account?upgrade_success=false`
+	});
+
+	response.redirect(303, session.url);
+});
+
+router.post('/stripe-wh', express.raw({ type: 'application/json' }), async (request, response) => {
+	const stripeSignature = request.headers['stripe-signature'];
+	let event;
+
+	try {
+		event = stripe.webhooks.constructEvent(request.body, stripeSignature, config.stripe.webhook_secret);
+	} catch (err) {
+		return response.status(400).send(`Webhook Error: ${err.message}`);
+	}
+
+	if (event.type === 'customer.subscription.updated' || event.type === 'customer.subscription.deleted') {
+		const subscription = event.data.object;
+		const product = await stripe.products.retrieve(subscription.plan.product);
+		const customer = await stripe.customers.retrieve(subscription.customer);
+		const pid = Number(customer.metadata.pnid_pid);
+		const pnid = await database.PNID.findOne({ pid });
```

Comment:
If account got deleted, this would return undefined and will throw an error later in the code. That returns a 500 and will make stripe infinitely retry this event.

Also if customer doesn't have meta data, this would return first account in db and modify it (equivalent of `findOne({})`). Should probably add a undefined check.

Reply:
When you say "account" are you talking about a customer with Stripe or a PNID with us? Because a PNID would never be deleted, only banned at worst (which I will handle later, as I don't want people buying these tiers to essentially unban themselves)

Also a customer should _never_ not have metadata. The customers used here are always created with the metadata. The server checks if a customer can be found with the users metadata before sending them off the checkout, and if it can't find a customer with the metadata it uses that one and forces the checkout to use it. So that webhook will never fire with a customer without metadata

I can always add a check for it anyway, but I can absolutely guarantee it will never happen (besides Stripe having a major fuck up on their end, in which case I might be worried about other things)

Though either way, I'm not really sure what the best course of action would be here? It's not like I can notify the user that the webhook failed, and at this stage the payment has already been processed. So I'm not really sure how to even effectively handle an error like that which would let both us _and_ the user know an issue happened, or how to rectify it?

Reply:
> Because a PNID would never be deleted, only banned at worst
thats fine, but do you really want a webhook to infinite loop just because you removed one test account, or one account was somehow so corrupted that it needed to be removed?

the account db can be controlled from many places, you cant garuantee it will never make one mistake. and thus you should make it so such a small mistake wont cause major failures, just keep a small mistake small by accounting for these things.
its loosly linked data, so it can fuck up.

> Also a customer should never not have metadata
assuming you never add any other payment system. lets say you rent out game servers in the future, that would make this code infinite retry loop again.

again its loosly linked data, easy to get out of sync or used in a different manner. just add a simple check so a simple oversight doesnt become a major issue.

> Though either way, I'm not really sure what the best course of action would be here?
if there is no pnid attached to the customer, just ignore the event and mark as success. its not meant for this service or integration then.
if the PNID account doesn't exist anymore, just ignore the event and mark as success. its probably an event from the past before the account got yeeted.
both should probably log something in the console, just for debugging purposes.

> but I can absolutely guarantee it will never happen
all these cases are if something small accidently happened. its de-escalation of an issue.
an accidental delete shouldn't cause a backlog of failed events. neither should a typo when you were looking through the stripe customer list in the UI.

Reply:
> the account db can be controlled from many places, you cant garuantee it will never make one mistake. and thus you should make it so such a small mistake wont cause major failures, just keep a small mistake small by accounting for these things.

Calling this a small issue kinda downplays the issue. A PNID being _deleted_ is a major issue by itself. That should never occur, ever, and there's no system in place where that could happen save someone wiping the DB manually?

> assuming you never add any other payment system. lets say you rent out game servers in the future, that would make this code infinite retry loop again.
> 
> again its loosly linked data, easy to get out of sync or used in a different manner. just add a simple check so a simple oversight doesnt become a major issue.

I'm not sure I understand? If I ever rent out game servers like that, it will be done through the PNID. So the same Stripe customer account will end up being used? There isn't a time where checkout would happen with a Stripe customer account that doesn't have a PNID attached, since it will always force one to be made if it can't find one?

> if there is no pnid attached to the customer, just ignore the event and mark as success. its not meant for this service or integration then.
> if the PNID account doesn't exist anymore, just ignore the event and mark as success. its probably an event from the past before the account got yeeted.

the problem with these are that both of these situations should never happen, and should not be allowed. If at any point a subscription webhook fires where the customer account has no PNID, or if the linked PNID can't be found, those are _major_ issues. That's not something small that can be just ignored. Those aren't valid inputs, that's very bad. Custom accounts are are always force created a PNID attached, if one isn't there something major has fucked up. And a PNID should never ever be deleted entirely, if that's the case something major has fucked up

> footnote: if you still disagree, thats fine. im not gonna argue this further.
> just know that you're making it really easy for user-error to be a big issue if you dont add these safety checks

I'm not trying to argue, I'm just trying to make sure we are on the same page with everything here and how the checkout flow goes so that we can reach a reasonable resolution